### PR TITLE
Conversion to Typescript

### PR DESCRIPTION
**Pull request by Gladear**
Hello !

I've been trying to convert **Haunted** to Typescript, and finally reached a correct state. The API does not change at all, and the minified bundle size is almost unchanged (less that 50 more bytes with _terser_, with options `--mangle --compress --toplevel`).
However, Typescript only permit to generate file with an indentation of 4 spaces, which greatly increases the size of generated lib files.

The conversion made me discover and correct two issues with the new State API, which should probably be corrected even if this PR is not merged:
- in the `useReducer` hook, the state of the reducer is stored on the same variable as the state of the Hook,
- in the `useContext` hook, the check that the component was not _virtual_ was made on the hook itself, not on its state.

One major change was required to convert to Typescript:
The `makeScheduler` was replaced by a `BaseScheduler` class, which is overriden for each type of renderer (_component_ and _virtual_).
I think this change makes sense because it allows greater flexibility if someone ever wanted to create virtual components with another library.

### CODE REVIEW

**Comment by matthewp** [comment]
This looks excellent! Thank you very much.  Any chance you'd like to take on #120?  I ask because my TypeScript skills are super low, and I'd like to get #120 into minor. If we go ahead and merge this it might slow that down... unless you're willing to do it.

It should be super easy, it's essentially the same as useEffect but runs immediately (synchronously) after render(templateResult).

**Comment by Gladear** [comment]
I'll definitely take a look into it, but before this is merged, I'd like someone to take a look, because I definitely don't consider my self a Typescript expert ! So if you know people that could review it, I'd be glad to hear there thought !

**Comment by chase-moskal** [comment]
@Gladear — hey, great work!

i did a quick review, and while i didn't have time to really delve into the details of the typings, i do have one recommendation about preserving the file history for the converted files

i made a quick three minute video about it for you, please check it out:

- https://www.twitch.tv/videos/476894384

let me know what you think, and let me know if you need any help understanding what i mean

cheers!

:wave: Chase

**Comment by matthewp** [comment]
Thanks @chase-moskal. Maybe someone else like @askbeka can chime in?

**Comment by matthewp** [comment]
Ok, watched the video, very good idea @chase-moskal! I also saw the commit you had there at the top, something you're working on? :D

**Comment by askbeka** [comment]
Cool. Will be happy to.
Will take a look today ( I am in CEST)

**Comment by Gladear** [comment]
Hi @chase-moskal,
I had not though about the files history, thanks for the tip and the video !

I almost forgot an important point.
There was actually a breaking change. I could not find a way to make the `symbolFor` function return a `unique symbol`, which allows to use the variables as properties for classes.
This makes a polyfill for `Symbol` mandatory for older browsers.
So if anyone has a workaround, it would be very nice from them to share it :blush:

File: src/core.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { makeComponent } from './component';
+import { makeContext } from './create-context';
+
+type GenericRenderer = (this: unknown, ...args: any[]) => unknown | void;
+type RenderFunction = (result: unknown, container: DocumentFragment | Element) => void;
+
+function haunted({ render }: { render: RenderFunction }) {
```
Comment (askbeka) [inline src/core.ts]: Can be better to expose options interface, making it easier to enxtend it later, as:

```ts
export interface Options {
  render: RenderFunction;
}

function haunted({ render }: Options)
```

Does it make sense?

**Comment by askbeka** [comment]
Overall looks good, except the point @chase-moskal addressed with files not preserving their origin, which makes it hard to review what has changed too.

Will be nice to to also set explicit return types on functions.
which makes it hard to break accidentally the API.

**Comment by Gladear** [comment]
It seems the files history is incorrectly displayed on GitHub, but using `git log --follow -- src/[file]`, or VS Code's _Git History_,  I can see the full history of the files, even before renaming them.

**Comment by matthewp** [comment]
@Gladear Awesome, let me know when you think this is ready for final review. Thanks so much for your hard work!

**Comment by Gladear** [comment]
Well, thank you @matthewp for your awesome library !

I think it might be ready for review. I'm currently working on [Haunted Router](https://github.com/Gladear/haunted-router), and the types seem to work pretty well to make a hook library. I also have a local test project to verify that the changes work well on the user side, and everything seem fine :blush:

**Comment by matthewp** [comment]
@chase-moskal is the history what you would expect now?

**Comment by chase-moskal** [comment]
@matthewp, @Gladear — yes, i have confirmed that the history is fixed, it looks great!

the fixed history makes the typescript conversion in commit 34c6277691c325a61eeb9f56e72c7eea9fa4ba41 very easy to review, and after a quick skim, it looks great to me :+1:

a couple of notes
- i'm extremely pleased to find every individual es module is now available in `lib` :+1:
- i noticed the `.d.ts` files are in the `typings` directory
  - in the past i've placed `.d.ts` files next to each corresponding `.js` in the destination folder (`lib`)
  - the benefit would be that direct imports like `import {haunted} from "haunted/lib/core.js"` would automatically find the sibling `.d.ts`, which should improve the situation for imports pointing directly into `lib`
  - i think simply removing the `declarationDir` config line from `tsconfig.json` would accomplish this

i haven't run the tests or exercised any of these changes, but from what i can see in gitk, this looks like great work :+1:

cheers

:wave: Chase

**Comment by matthewp** [comment]
I'll review this in the morning (provided I don't sleep in)

**Comment by matthewp** [comment]
Reviewed it and it looks good to me. I admit that not understanding the more advanced parts of the TypeScript definitions is a little bit intimidating to me, but I guess this means I'll just have to learn it now. Thanks to @chase-moskal and @askbeka for reviewing.

Merging! Big thanks to @Gladear for the effort on this, it was a lot of work, but I think it's going to make a lot of people more happy in the future.

**Pull request merged**